api/api-agents-v5/src/main/java/com/thoughtworks/go/apiv5/agents/representers/EnvironmentsRepresenter.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright 2019 ThoughtWorks, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.thoughtworks.go.apiv5.agents.representers;
+
+import com.thoughtworks.go.api.base.OutputListWriter;
+import com.thoughtworks.go.config.EnvironmentConfig;
+import com.thoughtworks.go.util.comparator.AlphaAsciiComparator;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.stream.Collectors;
+
+public class EnvironmentsRepresenter {
+    public static void toJSON(OutputListWriter writer, Collection<EnvironmentConfig> environments) {
+        List<EnvironmentConfig> sortedEnvironmentConfigs = environments.stream()
+                .sorted(new EnvironmentConfigComparator())
+                .collect(Collectors.toList());
+
+        for (EnvironmentConfig environment : sortedEnvironmentConfigs) {
+            writer.addChild((childWriter) -> {
+                childWriter.add("name", environment.name());
+                childWriter.add("associated_from_config_repo", !environment.isLocal());
```

Comment:
@akshaydewan -- On dashboard API response, we've defined `from_config_repo` field. can this also be called `from_config_repo`?

Secondly, we were thinking of sending the `origin` object as part of the API response, stating where the config entity defined.
The benefit of sending origin object is, we can specifically say, this `environment -> to -> agent` association has been defined in `example-foo` config repo. And have a link to the config-repo-spa as well for further details.

A similar issue for dashboard pipelines defined in config repo is listed [here](https://github.com/gocd/gocd/issues/4403).

**Note:** This doesn't need to happen now. can be done later too (at an overhead of bumping up API version) :)

Reply:
The tricky bit here is that we want to know *association* between an agent and an Environment. The Environment may be defined in a Config Repo, but it's association with an Agent can be defined from the UI (which will end up in XML). [See [EnvironmentConfig.isLocal()](https://github.com/gocd/gocd/blob/58886bb2b85294bb94e805651bb90ca75fae330f/config/config-api/src/main/java/com/thoughtworks/go/config/EnvironmentConfig.java#L95) vs [EnvironmentConfig.containsAgentRemotely(uuid)](https://github.com/gocd/gocd/blob/58886bb2b85294bb94e805651bb90ca75fae330f/config/config-api/src/main/java/com/thoughtworks/go/config/EnvironmentConfig.java#L105) ).

To avoid this ambiguity, I would rather include the word "associated". 

Regarding `origin`, we have the same ambiguity - is the Environment's origin, or the Environment-Agent relationship?

In case of the Dashboard, it is simpler because we need to know the origin of the Pipeline.

Reply:
If the association is in the XML, won't the environment tag be in the XML and won't the environment then be considered "local" and not remote?

Reply:
@arvindsv In such cases, the Environment becomes a `MergeEnvironmentConfig` object and it still returns `isLocal` as false.

Reply:
@akshaydewan I see. Ok.

Reply:
@akshaydewan @arvindsv  -- Thanks for the explainations. 

Though, on the enviornments, to resolve the ambiguity of entity-association and entity-definition, we were thinking of something like https://github.com/gocd/gocd/issues/3301#issuecomment-341654482

Wouldn't this apply here too?

We still can name the field `"association"` but can we make it return origin? If not, I am happy with whatever we have :)

Reply:
@GaneshSPatil and I had a chat about this. The way we reason about Environments is that the configuration can be `partial` - with some parts defined in the Config Repo, and other parts in XML. So each piece of the  environment configuration can have an `origin`. Ultimately the Environment config object becomes a `MergeEnvironmentConfig`.

With this reasoning, it would make sense to add an `origin` field in the API response. I can make the change in this PR itself to avoid another version bump.